Bobby requirement

petemeehan opened the pull request.

petemeehan committed:
used new "const" keyword for constants

compatible with IE11

petemeehan committed:
Code now runs in browser instead

Added monzo.html which runs the code, and test-data.js, to hold the
probability function code for generating the test data
Output only to console at this stage

petemeehan committed:
Wrapped scripts in scoping function to avoid global variables in window namespace

Rewrote code so that monzo.js now only provides functions, which is
executed separately (in this case in the html file).
Internal helper functions and variables are not visible outside of
script, only those functions which are hoisted to window object are
visible

petemeehan committed:
Added user interaction on web page

petemeehan committed:
Removed nested iterator

petemeehan committed:
Updated comments

petemeehan committed:
New private function to get random index of array

As it is used in multiple places

petemeehan committed:
Tidied small helper function

petemeehan committed:
Goal, pledge values and test data now set by method parameters

So can be set externally as passed arguments in web page script,
instead of being set in monzo.js

petemeehan committed:
Recalculate button disabled until after first calculation

To prevent functions being called out of order

petemeehan committed:
Tidied up webpage script

Set dom elements to variables

petemeehan committed:
Reorganised code a lil bit

petemeehan committed:
Handle errors with error messages

Error messages appear if not enough pledges to reach goal, or if too
many pledges fail

petemeehan committed:
Variables now saved to/retrieved from local storage

petemeehan committed:
Added a lil stylin

cmck committed:
Merge branch 'master' into bobby-requirement

cmck commented on monzo.js:
```diff
@@ -1,133 +1,239 @@
+  // Check local storage for existing variable values, otherwise start over
+  // and initialise with passed arguments
+  if (
+    localStorage.getItem("goal") !== null
```
we can assume that if the `goal` variable is unavailable on localstorage, then all of local storage is inconsistent and unreliable. so just do this single check.
Also, rather than use if/else look at ternary conditionals. https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Conditional_Operator

petemeehan replied:
ok to keep if/else here if i find it more readable in this case?

cmck commented on monzo.js:
```diff
@@ -1,133 +1,239 @@
+  return newDatasetObject;
+}
+
+function removeObjectSubset(subset, superset) {
```
are subset and superset used correctly here? subset has no relation to superset in this function

petemeehan replied:
think its correct? maybe worded ambiguously tho. the "superset" is the object the subset is being removed from.
maybe i should replace the word "superset" with "object"

cmck commented on monzo.html:
```diff
@@ -0,0 +1,117 @@
+        };
+
+        resetButton.onclick = function () {
+          console.log("test");
```
remove.

petemeehan replied:
woops

cmck commented on monzo.js:
```diff
@@ -1,133 +1,239 @@
+/*---------------PUBLIC FUNCTIONS---------------*/
+
+monzo.getGoal = function () {
+  return Number(goal).toLocaleString("en")
```
this method is public, so what if it is called before initialisation of `goal` ?

petemeehan replied:
it would return "NaN", as would the other get methods. Is this a problem though? These methods need to be public, so not sure how i would prevent outside scripts running them before initilisation?

cmck commented on monzo.js:
```diff
@@ -1,133 +1,239 @@
+  }
+
+  while (runningTotal >= goal ) {
+    var randomId = ids.splice(getRandomIndex(ids), 1)[0];
```
don't declare variables inside the loop

petemeehan replied:
oops, good catch 🎣

cmck commented on monzo.js:
```diff
@@ -1,133 +1,239 @@
+  return arr;
+}
+
+function saveVariablesToLocalStorage() {
```
better to split this into individual setter functions for encapsulation etc, also you're not updating everything unneccessarily each time you call it

petemeehan replied:
so should have 8 individual setter functions, one for each variable?

petemeehan replied:
also does that mean i do not need getter/setter functions to retrieve local variables? because all my getter/setter functions will retrieve the stored variables instead?

petemeehan replied:
implemented change [in latest commit](https://github.com/apmeehan/monzo-crowdfund/pull/3/commits/ac11c4f4393c3bc63c2ad8e55bab44fb08809a21).
now acceptable?

petemeehan committed:
Fixed variable declaration inside loop

petemeehan committed:
Removed checking for existence of all variables

Only checking of one of them is required, since all are practically
initialised at once

petemeehan committed:
Minor review fixes

petemeehan committed:
Split "saveAllVariables()" into individual save and load functions

petemeehan committed:
Improved readability by adding whitespace to parentheses

petemeehan committed:
minor whitespace editing
